experimental/dds/tree2/src/feature-libraries/optional-field/optionalField.ts

```diff
@@ -131,120 +139,131 @@ export const optionalChangeRebaser: FieldChangeRebaser<OptionalChangeset> = {
-			const fieldChangeInfo = revisionMetadata.tryGetInfo(
-				revision ?? change.fieldChange?.revision,
-			);
-			const { childChanges } = change;
-			if (childChanges !== undefined) {
-				for (const [childId, childChange] of childChanges) {
-					const taggedChildChange = tagChange(childChange, revision);
-					if (childId === "self") {
-						// childChange refers to the node that existed at the start of `change`,
-						// Thus in the composition, it should be referred to by whatever deletes that node in the future, which is what
-						// currentChildNodeChanges tracks
-						currentChildNodeChanges.push(taggedChildChange);
-					} else {
-						addChildChange(childId, taggedChildChange);
-					}
-				}
+			if (
+				isInputContextEmpty === undefined &&
+				(change.moves.length > 0 || change.reservedDetachId !== undefined)
+			) {
+				// change includes sets to the field, thus the emptiness of it previously can be inferred.
+				// Note that this block isn't entered for child-only changes.
+				isInputContextEmpty = change.reservedDetachId !== undefined;
 			}
+			const withIntention = (id: RegisterId): RegisterId => {
+				if (id === "self") {
+					return id;
+				}
+				const intention = getIntention(id.revision ?? revision, revisionMetadata);
+				return { revision: intention, localId: id.localId };
+			};
 
-			if (change.fieldChange !== undefined) {
-				if (fieldChange === undefined) {
-					fieldChange = {
-						id: change.fieldChange.id,
-						revision: fieldChangeInfo?.revision,
-						wasEmpty: change.fieldChange.wasEmpty,
-					};
-				} else {
-					fieldChange.id = change.fieldChange.id;
-					fieldChange.revision = fieldChangeInfo?.revision;
+			if (change.build !== undefined) {
+				for (const { id, set } of change.build) {
+					builds.push({
+						id: {
+							revision: getIntention(id.revision ?? revision, revisionMetadata),
+							localId: id.localId,
+						},
+						set,
+					});
 				}
+			}
+			const nextSrcToDst = new ChildChangeMap<
+				[dst: RegisterId, target: "nodeTargeting" | "cellTargeting"]
+			>();
+			const nextDstToSrc = new ChildChangeMap<RegisterId>();
 
-				let hasMatchingPriorInverse = false;
-				const maybePriorInverse = changes.findIndex((c) => {
-					const cChangeInfo = revisionMetadata.tryGetInfo(
-						// Change c may be a composite, in which case we need to look the revision of the fieldChange
-						c.revision ?? c.change.fieldChange?.revision,
-					);
-					return (
-						(cChangeInfo?.rollbackOf !== undefined &&
-							cChangeInfo?.rollbackOf === fieldChangeInfo?.revision) ||
-						(cChangeInfo?.revision !== undefined &&
-							cChangeInfo?.revision === fieldChangeInfo?.rollbackOf)
+			if (change.reservedDetachId !== undefined) {
+				latestReservedDetachId = withIntention(change.reservedDetachId);
+			}
+
+			// Compose all the things that `change` moved.
+			for (const [src, dst, target] of change.moves) {
+				let originalSrc = current.dstToSrc.get(withIntention(src));
```

Comment:
Consider de-duplicating all the evaluations for `withIntention(src)` and `withIntention(dst)`.